### DISCUSSION TITLE
on_exit pass parameters

### DISCUSSION THREAD

**Discussion by bnewman-tech** [pytransitions/transitions#590] [Q&A]
Howdy, 

I was wondering if on the on_exit you can pass parameters. If you have any thoughts please let me know. 

class Model(object):
    def Send_Email(self, emails: list):
      print(emails)
      print("Emailing Notify")

**Working:** 
         "on_exit":[
            "Create_Form"
         ]
**Goal:** 
         "on_exit":[
            "Create_Form(["xy@xy.com","ad@ad.com"])"
         ]

**Comment by aleneum** [comment]
A copy/paste  of my [SO answer](https://stackoverflow.com/a/68201420/1617563). Does this answer your question?:

The section of `transitions`' documentation called [Passing Data](https://github.com/pytransitions/transitions#passing-data):

> ... you can pass any positional or keyword arguments directly to the trigger methods (created when you call add_transition()) [...] You can pass any number of arguments you like to the trigger. There is one important limitation to this approach: every callback function triggered by the state transition must be able to handle all of the arguments.

For your particular example this could look like this:


```python
from transitions import Machine

class Matter(object):
    def say_hello(self, param):
        print(f"hello, new state! Here is your param: {param}")

    # Every callback MUST be able to process possible callback parameters
    # If parameters are not needed, just use *args and **kwargs in the definition
    def say_goodbye(self, *args):
        print("goodbye, old state!")


lump = Matter()
machine = Machine(lump, states=[{'name': 'solid', 'on_exit': 'say_goodbye'},
                                'liquid',
                                {'name': 'gas', 'on_enter': 'say_hello'}],
                  transitions=[['sublimate', 'solid', 'gas']], initial='solid')

# pass param as arg
lump.sublimate(lump.state)
# or as kwarg
# lump.sublimate(param=lump.state)

```

There is also a second way to pass data by passing `send_event=True` in the `Machine` constructor. This will change the way how `transitions` passes trigger parameters to callbacks:

> If you set send_event=True at Machine initialization, all arguments to the triggers will be wrapped in an EventData instance and passed on to every callback. (The EventData object also maintains internal references to the source state, model, transition, machine, and trigger associated with the event, in case you need to access these for anything.)

This might be more suitable for your use case since an `EventData` object also contains information about the executed transition which contains the name of the source state:

```python
from transitions import Machine, EventData

class Matter(object):
    def say_hello(self, event: EventData):
        print(f"hello, new state! Here is your param: {event.kwargs['param']}. "
              f"I came here from state '{event.transition.source}'.")

    def say_goodbye(self, event):
        print("goodbye, old state!")


lump = Matter()
machine = Machine(lump, states=[{'name': 'solid', 'on_exit': 'say_goodbye'},
                                'liquid',
                                {'name': 'gas', 'on_enter': 'say_hello'}],
                  transitions=[['sublimate', 'solid', 'gas']], initial='solid', send_event=True)

lump.sublimate(param=42)
```
